### PR TITLE
Enable quant op to share quantization parameter between input and ouput

### PR DESCRIPTION
**Pull request by yufenglee**
**Description**: Describe your changes.
Quantization parameters for some ops input and output need to be same, like Gather, Transpose, Split. This PR adds support for sharing quantization parameters between input and output.

### CODE REVIEW

**Review by jchen351 — APPROVED** [review]

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 1]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
         assert end == len(self.graph().node), "Graph is not a DAG"
         self.graph().ClearField("node")
         self.graph().node.extend(sorted_nodes)
+
+    def clean_initializers(self):
+        return ONNXModel.clean_initializers_helper(self.graph(), self.model)
+
+    @staticmethod
+    def clean_initializers_helper(graph, model):
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:376]: It is required that this function be a staticmethod? If not prefer making it a private function in the module. (prepend `_` to its name and move out from the class)

 https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 2]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
...
+        """
+        Clean unused initializers including which is caused by quantizing the model.
+            return cleaned graph, and list of tensor names from this graph and all its subgraphes
+            that can not be found in this graph and its subgraphes
+        """
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:381]:
```suggestion
        """Clean unused initializers including which is caused by quantizing the model.
        
        Returns:
            A cleaned graph, and a list of tensor names from this graph and all its subgraphs
            that cannot be found in this graph and its subgraphs
        """
```

nit: formatting, wording

File: onnxruntime/python/tools/quantization/onnx_model.py (hunk 1 above)
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:376]: Prefer adding type annotations for new functions for clarity (for people and type checkers)

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 3]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
...
+        requesting_tensor_names = {}
+        requesting_tensor_names.update(
+            {input_name: 1 for node in graph.node for input_name in node.input if input_name}
+        )
+        requesting_tensor_names.update({g_out.name: 1 for g_out in graph.output if g_out.name})
+
+        new_nodes = []
+        for node in graph.node:
+            node_2_add = node
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:390]: naming: Is there a name for `node_2_add` that conveys what it is better to readers?

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 4]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
...
+            graph_attrs = [
+                attr
+                for attr in node.attribute
+                if attr.type == onnx.AttributeProto.GRAPH or attr.type == onnx.AttributeProto.GRAPHS
+            ]
+            if len(graph_attrs) > 0:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:396]:
```suggestion
            if graph_attrs:
```

pythonic checks

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 5]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
...
+                kwargs = {}
+                for attr in node.attribute:
+                    kv = {}
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:399]: readability: Avoid abbreviations - they add cognitive load to readers.

> Do not use abbreviations that are ambiguous or unfamiliar to readers outside the project, and do not abbreviate by deleting letters within a word.

https://google.github.io/styleguide/pyguide.html#316-naming
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 6]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
...
+                    if attr.type == onnx.AttributeProto.GRAPH:
+                        (
+                            cleaned_sub_graph,
+                            sub_requesting_tensor_names,
+                        ) = ONNXModel.clean_initializers_helper(attr.g, model)
+                        kv = {attr.name: cleaned_sub_graph}
+                        requesting_tensor_names.update({gn: 1 for gn in sub_requesting_tensor_names})
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:406]: readability: Avoid abbreviations - they add cognitive load to readers.

> Do not use abbreviations that are ambiguous or unfamiliar to readers outside the project, and do not abbreviate by deleting letters within a word.

https://google.github.io/styleguide/pyguide.html#316-naming
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 7]
```diff
@@ -369,3 +368,93 @@ def topological_sort(self):
...
+                    kwargs.update(kv)
+                node_2_add = onnx.helper.make_node(node.op_type, node.input, node.output, name=node.name, **kwargs)
+            new_nodes.extend([node_2_add])
+
+        graph.ClearField("node")
+        graph.node.extend(new_nodes)
+
+        generated_names = {}
+        generated_names.update({output_name: 1 for node in graph.node for output_name in node.output if output_name})
+        for gn in generated_names:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:428]: readability: Avoid abbreviations - they add cognitive load to readers.

> Do not use abbreviations that are ambiguous or unfamiliar to readers outside the project, and do not abbreviate by deleting letters within a word.

https://google.github.io/styleguide/pyguide.html#316-naming
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: onnxruntime/python/tools/quantization/operators/gather.py [hunk 8]
```diff
@@ -44,9 +42,23 @@ def quantize(self):
...
-        gather_original_output = node.output[0]
         node.output[0] = gather_new_output
         node.input[0] = quantized_input_names[0]
         nodes.append(node)
 
         self.quantizer.new_nodes += nodes
+
+
+class QDQGather(QDQOperatorBase):
+    def __init__(self, onnx_quantizer, onnx_node):
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/operators/gather.py:53]: Prefer adding type annotations to new code for clarity and correctness

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 9]
```diff
@@ -4,40 +4,43 @@
...
+
+class QDQTensorQuantInfo:
+    def __init__(self, tensor_type=QDQQuantTensorType.NORMAL, quant_para_provider=None, axis=None):
+        self.tensor_type = tensor_type
+        self.quant_para_provider = quant_para_provider
+        self.axis = axis
+
+
 class QDQQuantizer(ONNXQuantizer):
     def __init__(
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:45]: not part of this pr, but: W0102: Dangerous default value {} as argument (dangerous-default-value). Use None instead

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 10]
```diff
@@ -69,9 +72,9 @@ def __init__(
             op_types_to_quantize,
             extra_options,
         )
-        self.tensors_to_quantize = []
-        self.tensors_to_quantize_per_channel = []
+        self.tensors_to_quantize = dict()
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:75]: R1735: Consider using {} instead of dict() (use-dict-literal)

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 11]
```diff
@@ -107,34 +110,53 @@ def __init__(
             else extra_options["QDQOpTypePerChannelSupportToAxis"]
         )
 
-    def quantize_tensor(self, tensor_name):
+    def _is_tensor_quantizable(self, tensor_name):
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:113]:
```suggestion
    def _is_tensor_quantizable(self, tensor_name: str) -> bool:
```

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 12]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
-                        self.model.replace_node_input(node, tensor_name, dq_output)
+    def _add_qdq_pair_for_activation(self, tensor_name, scale_name, zp_name):
+        if (
+            self.dedicated_qdq_pair
+            and tensor_name in self.tensor_to_its_receiving_nodes
+            and len(self.tensor_to_its_receiving_nodes[tensor_name]) > 1
+        ):
+            num_dedicated_qdq_pair = len(self.tensor_to_its_receiving_nodes[tensor_name])
+            for i in range(num_dedicated_qdq_pair):
+                postfix = "_" + str(i + 1)
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:277]:
```suggestion
                postfix = f"_{i + 1}"
```

nit

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 13]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
-                    q_output = tensor_name + "_QuantizeLinear"
-                    dq_input = q_output
-                    dq_output = tensor_name + "_DequantizeLinear"
-                    if self.model.is_graph_output(tensor_name):
-                        q_input = tensor_name + "_QuantizeLinearInput"
-                        dq_output = tensor_name
-                        self.model.replace_output_of_all_nodes(tensor_name, q_input)
+                    data_found, scale_name, zp_name, _, _ = self._get_quantization_params(tensor_name)
+
+                    if data_found == False:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:345]: `if not data_found`?

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 14]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
+                        raise ValueError(
+                            "Quantization parameters are not specified for param {}. "
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:347]: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
Raw output

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 15]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
+                            )
+                        )
+
+                    self._add_qdq_pair_for_activation(tensor_name, scale_name, zp_name)
+
+                del self.tensors_to_quantize[tensor_name]
+
+    def _quantize_sharing_param_tensors(self):
+        while self.tensors_to_quantize:
+            has_update = False
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:359]: nit: naming: `updated`, `has_updated`

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 16]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
-                    quantized_value = QuantizedValue(
-                        tensor_name,
-                        dq_output,
-                        scale_name,
-                        zp_name,
-                        QuantizedValueType.Input,
-                    )
-                    self.quantized_value_map[tensor_name] = quantized_value
+            if not has_update:
+                raise ValueError("There is acyclic dependence in quantization parameter shared mode")
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:379]: Just making sure: do you mean `a cyclic` dependency (a cycle)?

Comment (justinchuby) [reply onnxruntime/python/tools/quantization/qdq_quantizer.py:379]: It'd be helpful to also include an actionable suggestion in the error message

File: onnxruntime/python/tools/quantization/quant_utils.py [hunk 17]
```diff
@@ -549,3 +551,39 @@ def clone_model_with_shape_infer(model):
...
         cloned_model = save_and_reload_model(model)
     return cloned_model
+
+
+def tensor_proto_to_array(initializer):
+    if initializer.data_type == onnx_proto.TensorProto.FLOAT:
+        weights = onnx.numpy_helper.to_array(initializer)
+    else:
+        raise ValueError(
+            "Only float type quantization is supported. Weights {} is {}. ".format(
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/quant_utils.py:561]: C0209: Formatting a regular string which could be a f-string (consider-using-f-string)
Raw output

File: onnxruntime/python/tools/quantization/quant_utils.py [hunk 18]
```diff
@@ -549,3 +551,39 @@ def clone_model_with_shape_infer(model):
     else:
         cloned_model = save_and_reload_model(model)
     return cloned_model
+
+
+def tensor_proto_to_array(initializer):
+    if initializer.data_type == onnx_proto.TensorProto.FLOAT:
+        weights = onnx.numpy_helper.to_array(initializer)
+    else:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/quant_utils.py:559]: I would remove the (redundant) else branch and raise directly.

File: onnxruntime/test/python/quantization/test_qdq.py [hunk 19]
```diff
@@ -208,23 +207,23 @@ def test_qdq_extra_options_2(self):
         # QDQ pair shoud not be added to node's output
         for node in quantizer.model.nodes():
             if node.name == "MatMul1":
-                self.assertTrue("T_DequantizeLinear_1" in node.input)
+                self.assertTrue("T_DequantizeLinear_Output_1" in node.input)
```
Comment (justinchuby) [inline onnxruntime/test/python/quantization/test_qdq.py:210]: Prefer https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertNotIn for clarity when there is an error

File: onnxruntime/test/python/quantization/test_qdq.py [hunk 20]
```diff
@@ -208,23 +207,23 @@ def test_qdq_extra_options_2(self):
...
             if node.name == "Add":
                 for input in node.input:
-                    self.assertTrue("DequantizeLinear" not in input)
+                    self.assertTrue("L_DequantizeLinear_Output" not in input)
 
             # QDQ pair shoud not be added to MatMul's output
             if node.op_type == "QuantizeLinear":
                 self.assertTrue(
                     node.input[0]
                     not in [
```
Comment (justinchuby) [inline onnxruntime/test/python/quantization/test_qdq.py:223]: Consider using a set for checking membership?

`... in {"", ...}`

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 21]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
+                tensor_provider_name = tensor_info.quant_para_provider
+                if tensor_provider_name in self.quantized_value_map:
+                    del self.tensors_to_quantize[tensor_name]
+
+                    quantized_value = self.quantized_value_map[tensor_provider_name]
+                    # Quantize the input
+                    initializer = find_by_name(tensor_name, self.model.initializer())
+                    if initializer is not None:
+                        raise ValueError("Quantization parameter shared mode is not supported for weight yet")
                     else:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:370]: R1720: Unnecessary "else" after "raise", remove the "else" and de-indent the code inside it (no-else-raise)
Raw output

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 22]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
-                        quant_node_name,
-                    )
-                    dequant_node = onnx.helper.make_node(
-                        "DequantizeLinear",
-                        [dq_input, scale_name, zp_name],
-                        [dq_output],
-                        dequant_node_name,
-                    )
-                    self.model.add_nodes([qlinear_node, dequant_node])
+                        scale_name = quantized_value.scale_name
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:371]: W0612: Unused variable 'scale_name' (unused-variable)
Raw output

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 23]
```diff
@@ -181,137 +203,182 @@ def try_replacing_upstream_output(self, upstream_output_name, output_name):
...
+                        zp_name = quantized_value.zp_name
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:372]: W0612: Unused variable 'zp_name' (unused-variable)
Raw output

File: onnxruntime/python/tools/quantization/qdq_quantizer.py (hunk 11 above)
Comment (yufenglee) [reply onnxruntime/python/tools/quantization/qdq_quantizer.py:113]: will have a separate PR to update the type annotations for all functions under quantization.

File: onnxruntime/python/tools/quantization/onnx_model.py [hunk 24]
```diff
@@ -124,9 +203,11 @@ def get_parent(self, node, idx, output_name_to_node=None):
         return output_name_to_node[input]
 
     def find_node_by_name(self, node_name, new_nodes_list, graph):
-        """
-        Find out if a node exists in a graph or a node is in the
-        new set of nodes created during quantization. Return the node found.
+        """Find out if a node exists in a graph or a node is in the
+        new set of nodes created during quantization.
+
+        Return:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/onnx_model.py:209]: nit: https://google.github.io/styleguide/pyguide.html#doc-function-returns

```suggestion
        Returns:
```

File: onnxruntime/python/tools/quantization/qdq_quantizer.py [hunk 25]
```diff
@@ -365,18 +362,9 @@ def _quantize_sharing_param_tensors(self):
                     quantized_value = self.quantized_value_map[tensor_provider_name]
                     # Quantize the input
                     initializer = find_by_name(tensor_name, self.model.initializer())
-                    if initializer is not None:
+                    if initializer:
```
Comment (justinchuby) [inline onnxruntime/python/tools/quantization/qdq_quantizer.py:365]: nit: explicit `is not None` is generally good

https://google.github.io/styleguide/pyguide.html#2144-decision

**Review by justinchuby — APPROVED** [review]
thanks!

**Review by justinchuby — APPROVED** [review]
